### PR TITLE
Extend max rules per rule group limit to be supported per namespace

### PR DESCRIPTION
**Pull request by gotjosh**
#### What this PR does

Introduces a new limit named `ruler_max_rules_per_rule_group_by_namespace` which allows one to specify which namespaces do we want to be different to have a different limit in terms of rules per group.

With this change, we now have a hierarchy for this limit as follows:

1. Per tenant limit for the given namespace.
2. Default limit for the given namespace.
3. Per tenant limit set by RulerMaxRulesPerRuleGroup
4. Default limit set by RulerMaxRulesPerRuleGroup

This is part 1 of the change, and I need to extend the same level of support to the other limit related to rule groups of `MaxRuleGroupsPerTenant`.

**Marking it as a draft as I still have some work to do around the documentation and experimental guarantees** 

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/util/validation/limits_map.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+package validation
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"gopkg.in/yaml.v3"
+)
+
+// LimitsMap is a generic map that can hold either float64 or int64 as values.
```
Comment (gotjosh) [inline pkg/util/validation/limits_map.go:12]: I need to use this 3 times total (1 for notifications, 1 for rules per group and 1 for rule groups by tenant) - I thought a generic was warranted but let me know if not.

File: pkg/mimirtool/config/inspect.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ var (
 type InspectedEntryFactory func() *InspectedEntry
 
 // InspectedEntry is the structure that holds a configuration block or a single configuration parameters.
-// Blocks contain other other InspectedEntries.
+// Blocks contain other InspectedEntries.
```
Comment (gotjosh) [inline pkg/mimirtool/config/inspect.go:32]: drive-by typo fix.

File: pkg/util/validation/limits_map.go (hunk 1 above)
Comment (gotjosh) [reply pkg/util/validation/limits_map.go:12]: I'm happy to also move this change to a separate PR if it makes your life easier but thought the refactor w/o context didn't made much sense.

File: pkg/util/validation/limits_test.go [hunk 3]
```diff
@@ -586,6 +587,193 @@ testuser:
 	}
 }
 
+func TestRulerMaxRulesPerRuleGroupLimits(t *testing.T) {
+	tc := []struct {
+		name      string
+		inputYAML string
+		expected  struct {
```
Comment (pstibrany) [inline pkg/util/validation/limits_test.go:594]: nit: can we use `expectedLimit`, `expectedNamespace`, and avoid repeating struct definition in every test?

File: pkg/util/validation/limits_test.go [hunk 4]
```diff
@@ -586,6 +587,193 @@ testuser:
...
+    mynamespace: 500
+`
+
+	tc := []struct {
+		name           string
+		overrides      string
+		inputNamespace string
+		expectedLimit  int
+	}{
+		{
```
Comment (pstibrany) [inline pkg/util/validation/limits_test.go:689]: super-nit: if order is not important, `map[string]testCase` is preferrable, to avoid duplicate names.

File: pkg/util/validation/limits.go [hunk 5]
```diff
@@ -428,13 +434,21 @@ func (l *Limits) validate() error {
...
-	for k, v := range defaults {
-		l.NotificationRateLimitPerIntegration[k] = v
+func (l *Limits) copyNotificationIntegrationLimits(defaults LimitsMap[float64]) {
+	l.NotificationRateLimitPerIntegration = NotificationRateLimitMap()
+	for k, v := range defaults.data {
+		l.NotificationRateLimitPerIntegration.data[k] = v
 	}
 }
 
+func (l *Limits) copyMaxRulesPerRuleGroupByNamespaceLimits(defaults LimitsMap[int64]) {
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/util/validation/limits.go]: Good shout, done in https://github.com/grafana/mimir/pull/8378/commits/6f2c6452ca24c4607ec5c75ac9b608fba20f33f2

File: pkg/util/validation/limits_test.go (hunk 3 above)
Comment (gotjosh) [reply pkg/util/validation/limits_test.go]: Done in https://github.com/grafana/mimir/pull/8378/commits/426289646b3ba7479d456e82f20b74c3432a9bd2

File: pkg/util/validation/limits_test.go (hunk 4 above)
Comment (gotjosh) [reply pkg/util/validation/limits_test.go]: Done in https://github.com/grafana/mimir/pull/8378/commits/426289646b3ba7479d456e82f20b74c3432a9bd2 - I actually changed all of the tests that followed the above pattern.

File: pkg/util/validation/notifications_limit_flag.go [hunk 6]
```diff
@@ -6,57 +6,24 @@
...
-}
-
-type NotificationRateLimitMap map[string]float64
-
-// String implements flag.Value
-func (m NotificationRateLimitMap) String() string {
-	out, err := json.Marshal(map[string]float64(m))
-	if err != nil {
-		return fmt.Sprintf("failed to marshal: %v", err)
+var integrationsValidation = func(k string, _ float64) error {
```
Comment (colega) [inline pkg/util/validation/notifications_limit_flag.go:14]:
```suggestion
func validateIntegrationLimit(k string, _ float64) error {
```

File: pkg/util/validation/limits.go [hunk 7]
```diff
@@ -305,6 +306,10 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&l.RulerRecordingRulesEvaluationEnabled, "ruler.recording-rules-evaluation-enabled", true, "Controls whether recording rules evaluation is enabled. This configuration option can be used to forcefully disable recording rules evaluation on a per-tenant basis.")
 	f.BoolVar(&l.RulerAlertingRulesEvaluationEnabled, "ruler.alerting-rules-evaluation-enabled", true, "Controls whether alerting rules evaluation is enabled. This configuration option can be used to forcefully disable alerting rules evaluation on a per-tenant basis.")
 	f.BoolVar(&l.RulerSyncRulesOnChangesEnabled, "ruler.sync-rules-on-changes-enabled", true, "True to enable a re-sync of the configured rule groups as soon as they're changed via ruler's config API. This re-sync is in addition of the periodic syncing. When enabled, it may take up to few tens of seconds before a configuration change triggers the re-sync.")
+	if l.RulerMaxRulesPerRuleGroupByNamespace.data == nil {
```
Comment (colega) [inline pkg/util/validation/limits.go:309]: Why wouldn't it be nil?

Also, maybe we could have a helper method like `isInitialized()`? Or maybe even store a pointer to `LimitsMap` instead of having a zero-struct and having to check an unexported field.

File: pkg/util/validation/limits_map.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+func (m LimitsMap[T]) updateMap(newMap map[string]T) error {
+	for k, v := range newMap {
+		if m.validator != nil {
+			if err := m.validator(k, v); err != nil {
+				return err
+			}
+		}
+
+		m.data[k] = v
+	}
```
Comment (colega) [inline pkg/util/validation/limits_map.go:70]: I think we should validate all fields first, then update.

File: pkg/util/validation/limits.go [hunk 9]
```diff
@@ -831,8 +831,19 @@ func (o *Overrides) RulerTenantShardSize(userID string) int {
...
+// This limit is special. Limits are returned in the following order:
+// 1. Per tenant limit for the given namespace.
+// 2. Default limit for the given namespace.
+// 3. Per tenant limit set by RulerMaxRulesPerRuleGroup
+// 4. Default limit set by RulerMaxRulesPerRuleGroup
+func (o *Overrides) RulerMaxRulesPerRuleGroup(userID, namespace string) int {
+	u := o.getOverridesForUser(userID)
+
+	if namespaceLimit, ok := u.RulerMaxRulesPerRuleGroupByNamespace.data[namespace]; ok {
+		return int(namespaceLimit)
```
Comment (colega) [inline pkg/util/validation/limits.go:843]: Why would we define this map as `LimitsMap[int64]` if we're going to cast to an `int` later? Can we just define it as `LimitsMap[int]` from the beginning?

Comment (gotjosh) [reply pkg/util/validation/limits.go:843]: Yeah, I agree with you: https://github.com/grafana/mimir/pull/8378/commits/6c9a3bff174498dc8bd35764bbf024d8e4362617#diff-1e0a07c0c1ea87ab5a582d79c587a73a9b8e0680d89c61d5c94be106823a4a4e

File: pkg/util/validation/notifications_limit_flag.go (hunk 6 above)
Comment (gotjosh) [reply pkg/util/validation/notifications_limit_flag.go]: Done in https://github.com/grafana/mimir/pull/8378/commits/43f3a15f13e9d32b61e7d74638d746ef9b5ad3e6

File: pkg/util/validation/limits_map.go (hunk 8 above)
Comment (gotjosh) [reply pkg/util/validation/limits_map.go]: Good shout. I ported over the implementation we had before, but I agree with you that not allowing partial updates seems like a better approach - this is actually a breaking change (or a bug fix?) from the way the rate limit limit works so I'll have to add to the readme. Can you confirm that this assumption is correct?

Done in https://github.com/grafana/mimir/pull/8378/commits/e5440806807ea215073a31fac4bd4275ee57fedf

File: pkg/util/validation/limits.go (hunk 9 above)
Comment (gotjosh) [reply pkg/util/validation/limits.go]: Done in https://github.com/grafana/mimir/pull/8378/commits/7c6c7370e8c8c4c0355c39f6ea03bcfda93294a5

File: pkg/util/validation/limits.go [hunk 10]
```diff
@@ -178,13 +178,14 @@ type Limits struct {
...
+	RulerTenantShardSize                 int              `yaml:"ruler_tenant_shard_size" json:"ruler_tenant_shard_size"`
+	RulerMaxRulesPerRuleGroup            int              `yaml:"ruler_max_rules_per_rule_group" json:"ruler_max_rules_per_rule_group"`
+	RulerMaxRuleGroupsPerTenant          int              `yaml:"ruler_max_rule_groups_per_tenant" json:"ruler_max_rule_groups_per_tenant"`
+	RulerRecordingRulesEvaluationEnabled bool             `yaml:"ruler_recording_rules_evaluation_enabled" json:"ruler_recording_rules_evaluation_enabled" category:"experimental"`
+	RulerAlertingRulesEvaluationEnabled  bool             `yaml:"ruler_alerting_rules_evaluation_enabled" json:"ruler_alerting_rules_evaluation_enabled" category:"experimental"`
+	RulerSyncRulesOnChangesEnabled       bool             `yaml:"ruler_sync_rules_on_changes_enabled" json:"ruler_sync_rules_on_changes_enabled" category:"advanced"`
+	RulerMaxRulesPerRuleGroupByNamespace LimitsMap[int64] `yaml:"ruler_max_rules_per_rule_group_by_namespace" json:"ruler_max_rules_per_rule_group_by_namespace" category:"advanced"`
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/util/validation/limits.go]: I prefer the current naming approach, even if slightly different - if others feel strongly about it I'll rename it.

File: pkg/util/validation/limits.go (hunk 7 above)
Comment (gotjosh) [reply pkg/util/validation/limits.go]: > Also, maybe we could have a helper method like isInitialized()? 
Done in https://github.com/grafana/mimir/pull/8378/commits/a2e11637becfaa53cc97ba9a47f43cabe6bcb9ae

> Why wouldn't it be nil?
We wouldn't care if it's nil if you're specifying a value for the argument directly as it'll be marshalled from the JSON. When you're not, we care about having a default value so that it show up in the documentation. 

I am not sure if this is the answer to your question - if it isn't feel free to elaborate and I'll be happy to answer/fix your comment.

File: pkg/util/validation/limits_map.go (hunk 8 above)
Comment (colega) [reply pkg/util/validation/limits_map.go]: I think it's a minor bug that might not even be worth mentioning in the changelog. The fix lgtm.

**Review by colega — APPROVED** [review]
LGTM

File: pkg/util/validation/limits.go [hunk 11]
```diff
@@ -185,6 +185,7 @@ type Limits struct {
 	RulerRecordingRulesEvaluationEnabled bool           `yaml:"ruler_recording_rules_evaluation_enabled" json:"ruler_recording_rules_evaluation_enabled" category:"experimental"`
 	RulerAlertingRulesEvaluationEnabled  bool           `yaml:"ruler_alerting_rules_evaluation_enabled" json:"ruler_alerting_rules_evaluation_enabled" category:"experimental"`
 	RulerSyncRulesOnChangesEnabled       bool           `yaml:"ruler_sync_rules_on_changes_enabled" json:"ruler_sync_rules_on_changes_enabled" category:"advanced"`
+	RulerMaxRulesPerRuleGroupByNamespace LimitsMap[int] `yaml:"ruler_max_rules_per_rule_group_by_namespace" json:"ruler_max_rules_per_rule_group_by_namespace" category:"advanced"`
```
Comment (pracucci) [inline pkg/util/validation/limits.go:188]: I would mark it experimental first and list it among experimental features in `docs/.../about-versioning.md`. Once the entire work is done and we're happy with it, then we can switch it to advanced.

**Review by pracucci — APPROVED** [review]
Very nice design, good job! I left a nit.

File: pkg/util/validation/limits.go (hunk 11 above)
Comment (gotjosh) [reply pkg/util/validation/limits.go]: Done in https://github.com/grafana/mimir/pull/8378/commits/7e45e06d204213f3e7125ca5fd72022e43c50e08

When CI passes, I'll merge this.
